### PR TITLE
329 - Autocomplete mixin

### LINKED ISSUE #329: IdsAutoCompleteMixin: Add Mixin
Add a new autocomplete mixin MVP https://main-enterprise.demo.design.infor.com/components/autocomplete/list to the web component repo. 

Use new ids-listbox and extract some functionality thats used in dropdown (keyboard/accessibity/selection/template ect)

Try it on lookup and input first

**Phase 1 Features (MVP)**
- ajax
- used on lookup, input and search

**Phase 2 Features (MVP)**
- Most Anything else 

**Possible Internal Markup**
```js
<ids-input autocomplete="true">
</ids-input>
```

**Suggested Dependencies**
- ids-list-box https://github.com/infor-design/enterprise-wc/tree/main/src/components/ids-list-box
- ids-input
- custom keydownend event in idsevents

**Settings/Or Slots**
- async call
- template

**Design**
https://main-enterprise.demo.design.infor.com/components/autocomplete/example-index.html

**Events**
- selected/changed
- cleared

**Other**
Pass the following checklist ordered in priority (some may be left for later) https://github.com/infor-design/enterprise-wc/blob/main/doc/CHECKLIST.md

### PR DESCRIPTION
**Pull request by jmacaluso711**
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR adds the autocomplete mixin to `ids-input`

**Related github/jira issue (required)**:
closes #329 

**Steps necessary to review your pull request (required)**:
- Pull and run branch
- Go to http://localhost:4300/ids-input/autocomplete.html and start typing a state name in the input
- Make sure popup opens with results.
- Select a state by clicking. The input value should update.
- Select a state by navigating up/down with keyboard and press enter. The input value should update.

- Go to http://localhost:4300/ids-trigger-field
- Repeat the same steps from above in the input labeled "Autocomplete Trigger Field"

- Go to http://localhost:4300/ids-search-field
- Repeat the same steps from above in the input labeled "Autocomplete Search Field"

- Go to http://localhost:4300/ids-lookup/autocomplete.html
- Repeat the same steps from above. Search for description (i.e 101, 102, 103).

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

### CODE REVIEW

**Review by andriivarhanov — COMMENTED** [review]
looks good, everything works, some adjustments to add/discuss:
- Please add some events to autocomplete. Events in the original autocomplete beforepopulated, populated, listopen, listclose, selected. I’m thinking maybe `selected` and `cleared` like suggested in https://github.com/infor-design/enterprise-wc/issues/329
- Keyboard Escape should close popup
- Keyboard arrow up/down should cycle through the list. First item selected -> arrow up -> last item selected and vise versa
- Please add all autocomplete.html examples to component yaml files
ids-lookup: I think maybe use same `search-field` attribute instead of `field` for consistency
- If possible to add more tests to autocomplete to increase test coverage

File: src/components/ids-input/README.md [hunk 1]
```diff
@@ -91,15 +91,46 @@ Set up for displaying the show/hide password button:
...
 
+- `autocomplete` can be set to true to enable autocomplete functionality
 - `autoselect` {boolean} set auto select text on focus to input.
 - `bgTransparent` {boolean} set the transparent background to readonly input.
 - `clearable` {boolean} set (x) button to clear text on click/key to input. See [Ids Clearable Mixin](../../mixins/ids-clearable-mixin/README.md) for more information.
 - `clearableForced` {boolean} set (x) button to clear text on click/key to input, forced to be on readonly.
 - `colorVariant` {string} set the current color variant.
 - `compact` {boolean} sets the component to be compact mode.
 - `caps-lock` {boolean} sets whether the capslock indicator appears when caps lock is on.
+- `data` when autocomplete is enabled a instance of [IdsDatasource](../../core/README.md)]
```
Comment (tmcconechy) [inline src/components/ids-input/README.md:133]:
```suggestion
- `data` when autocomplete is enabled an instance of [IdsDatasource](../../core/README.md)]
```

File: src/core/ids-base.scss [hunk 2]
```diff
@@ -190,3 +190,7 @@ $transparent-white: rgba(255 255 255 / 0);
     @content;
   }
 }
+
+.highlight {
```
Comment (tmcconechy) [inline src/core/ids-base.scss:194]: We should keep this out of ids-base.scss -> i was working on cleaning this up on another PR as this class will end up in every component even if not needed.

You could do it as a mixin instead that way the class isnt directly added and put it in general-mixins? Or just put it where its needed in the component css

File: src/components/ids-input/demos/autocomplete.html [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  <ids-container role="main" padding="8" hidden>
+    <ids-theme-switcher mode="light" version="new"></ids-theme-switcher>
+    <ids-layout-grid auto="true" gap="md">
+      <ids-layout-grid-cell>
+
+        <ids-input
+          id="input-autocomplete-2"
+          placeholder="This input's label is visible"
+          size="md"
+          label="Autocomplete Input 2"
```
Comment (tmcconechy) [inline src/components/ids-input/demos/autocomplete.html:16]:
```suggestion
          label="Autocomplete Input (states)"
```

**Review by tmcconechy — CHANGES_REQUESTED** [review]
See a couple things worth considering.

A) One small behavior i noticed is if you 
- type a (any example)
- down arrow repeatedly
- doesn't scroll you cant see the items

B) Also if you go to mobile size (345px ish) the popup doesnt lineup.

Screen Shot 2022-04-12 at 9 42 16 AM

C) Also the under border is a different color... (minor)
Screen Shot 2022-04-12 at 9 40 52 AM

D) Also in dark mode the colors are a bit off
Screen Shot 2022-04-12 at 9 42 16 AM

**Review by tmcconechy — COMMENTED** [review]
All my points look addressed now. Thanks. Except for one about dark mode. I see i attached the wrong file. But it looks like this:

Screen Shot 2022-04-13 at 9 24 49 AM


One other thing i noticed is "No Results" is "hardcoded" in ids-autocomplete.js - should pull this out use Locale.translate() (ids-locale-mixin)

**Comment by tmcconechy** [comment]
Ok makes sense @jmacaluso711 can you make a separate issue? I can look as i was looking at something similar on that other PR, and i know that it did work when i made the dropdown originally. I think i may have broken this during TS or Webpack rework.

**Comment by tmcconechy** [comment]
@jmacaluso711 i fixed it on this PR https://github.com/infor-design/enterprise-wc/pull/647/commits/4715cd76fb35e046022fba4e008b76a536450da7#diff-65bc07cdd0a0e5e100943aa1cfd7d9007d73ea78e9decfa94b85e38f4c0ba1f4R37

Was just missing the connectedCallback (think i might have accidentally removed). So the theme mixin isnt registering
If you want to copy that in this branch for any reason feel free. But im willing to approve.

**Review by tmcconechy — APPROVED** [review]

**Pull request merged**